ccp/src/proof_queue.rs

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2024 Fluence DAO
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+use ccp_shared::proof::CCProof;
+use ccp_shared::proof::ProofIdx;
+use ccp_shared::types::CUID;
+use itertools::Itertools;
+
+// TODO it is not quite a queue; cache?
+#[derive(Debug, Default)]
+pub(crate) struct ProofQueue {
+    cu_proofs: HashMap<CUID, CuState>,
+}
+
+#[derive(Debug, Default)]
+struct CuState {
+    proofs: VecDeque<CCProof>,
+}
+
+impl CuState {
+    fn push_back(&mut self, proof: CCProof) {
+        self.proofs.push_back(proof)
+    }
+
+    fn len(&self) -> usize {
+        self.proofs.len()
+    }
+
+    fn try_get_batch(&mut self, batch_length: usize) -> Option<Vec<CCProof>> {
+        if self.proofs.len() >= batch_length {
+            // TODO is there anyting better than a VecDec drain?
+            // VecDec::split_off seems to produce a contiguous VecDec that can be turned into Vec at O(1),
+            // but it is not guaranteed
+            //
+            // you may think that iterator chunking would work here, but it is may be too greedy and one
+            // can lose items (i.e. proofs)
+            Some(self.proofs.drain(..batch_length).collect_vec())
```

Comment:
This is a destructive reading; actually, the reading should be non-destructive (using &iterators), but only `rewind_forward` on the next call should update the queue.  It applies to batch proofs and all proofs methods.